### ISSUE TITLE
Stream Processing with Apache Flink

### ISSUE THREAD

**Issue by Sunt-ing** [Sunt-ing/database-system-readings#112]
## Chapter 1

Event-Driven Applications
- Event-driven applications are stateful streaming applications that ingest event streams and process the events with application-specific business logic.
- Typical use cases for event-driven applications include:
  - Real-time recommendations (e.g., for recommending products while customers browse a retailer’s website)
  - Pattern detection or complex event processing (e.g., for fraud detection in credit card transactions)
  - Anomaly detection (e.g., to detect attempts to intrude a computer network)
- Event-driven applications are an evolution of microservices. They communicate via event logs instead of REST calls and hold application data as local state instead of writing it to and reading it from an external datastore, such as a relational database or key-value store. 

Table API is in LINQ style.

## Chapter 2
Modern stream processors, like Apache Flink, can offer latencies as low as a few milliseconds.

Processing-time windows introduce the lowest latency possible. Since you do not take into consideration late events and out-of-order events, a window simply needs to buffer up events and immediately trigger computation once the specified time length is reached. Thus, for applications where speed is more important than accuracy, processing time comes in handy.
Another case is when you need to periodically report results in real time, independently of their accuracy. An example application would be a real-time monitoring dashboard that displays event aggregates as they are received.
Finally, processing-time windows offer a faithful representation of the streams themselves, which might be a desirable property for some use cases. For instance, you might be interested in observing the stream and counting the number of events per second to detect outages. 
To recap, processing time offers low latency but results depend on the speed of processing and are not deterministic.

It is important to note that sometimes you can get stronger semantics with weaker guarantees. A common case is when a task performs idempotent operations, like maximum or minimum. In this case, you can achieve exactly-once semantics with at-least-once guarantees.

## Chapter 3. The Architecture of Apache Flink
Flink does not provide durable, distributed storage. Instead, it takes advantage of distributed filesystems like HDFS or object stores such as S3. For leader election in highly available setups, Flink depends on Apache ZooKeeper.
